Branch 201392512

mingxingtan opened the pull request.

Commit:
Minor python change to remove doing unnecessary work in resource variables

PiperOrigin-RevId: 200735157

tensorflower-gardener committed:
[TF:XLA] Validate the control flow structure in encapsulate_subgraphs_pass and encapsulate_tpu_computations_pass, in order to detect errors earlier.

PiperOrigin-RevId: 200735435

pavithrasv committed:
Increase gru_test test size

PiperOrigin-RevId: 200736300

hawkinsp committed:
[TF:XLA] Update comment on xla_compiler.h to match the code.

Make resource_var.h more widely visible and add comment about the correct lock acquisition order if locking multiple variables.

PiperOrigin-RevId: 200736416

tensorflower-gardener committed:
Fix a bug in dependency optimizer: Repeated inputs would not get converted to control inputs when converting nodes to NoOps.

PiperOrigin-RevId: 200740844

gunan committed:
Disable long running tests in fastbuild mode.

PiperOrigin-RevId: 200741660

pavithrasv committed:
Fix: DepthwiseConv2D fails when bias is enabled

PiperOrigin-RevId: 200742104

d0k committed:
[XLA] Switch PostOrder accessors to use std::vector instead of std::list.

std::list is just hilariously inefficient and the postorder list creation has
been rewritten not to not depend on splicing anymore so there's no need for the
list. While there remove the old unused postorder list creation code.
PiperOrigin-RevId: 200743677

meheffernan committed:
Add test of TOKEN primitive type which uses conditionals.

PiperOrigin-RevId: 200745718

tensorflower-gardener committed:
[XLA] Don't implement kCrossReplicaSum case in HloInstruction::IdenticalSlowPath.

PiperOrigin-RevId: 200746735

yk5 committed:
contrib.timeseries: sets the predictions dict in EstimatorSpec for evaluation op.

PiperOrigin-RevId: 200747192

k-w-w committed:
Refactor loader.load function into a class that splits the graph loading and variable restoration steps.

PiperOrigin-RevId: 200747752

kayzhu committed:
[TF2XLA] Remove the last unncessary host-to-device memcpy, and remove the
HostTensorToLiteral function completely to prevent potential future misuse of
unnecessary memcpy.

PiperOrigin-RevId: 200750664

Commit:
Support model parallelism in PER_HOST_V2 input pipeline.

PiperOrigin-RevId: 200751151

aaroey committed:
Add DeviceSet to SingleMachine, so we can use the OptimizeGraph() tool to call
tensorrt optimizer (which requires access to the Device) to create a
transformed GraphDef.

PiperOrigin-RevId: 200751174

Commit:
Broad refactoring (part 2): Introduce a module dedicated to AutoGraph-specific conversion logic: base converter classes, context objects, gensym. Largely, these are pulled out from impl and pyct.
This CL only adds the module - a future CL will replace existing implementations with these.

PiperOrigin-RevId: 200751782

tensorflower-gardener committed:
Reuse duplicated reference ops in optimized_ops.h

PiperOrigin-RevId: 200753184

tensorflower-gardener committed:
Fix segfault in ConstantFolding::MaterializeShapes when the first input to TensorArraySizeV3 is a Placeholder.

PiperOrigin-RevId: 200755274

tensorflower-gardener committed:
Provides a more fine-grained option for each thread to control
fork-join parallelism (e.g., Eigen/ThreadpoolDevice or Shard).

PiperOrigin-RevId: 200756626

hawkinsp committed:
Add some extra DebugString() functions to shape_inference. Currently unused, but they were useful while debugging.
Open visibility of the low level gen_resource_variables_ops to compiler tests.
Fix bug in shape function of TPUReplicateInput for resource variables ? MergeInputHandleShapesAndTypes does not report shape mismatches.

PiperOrigin-RevId: 200756762

tensorflower-gardener committed:
Set shapes and types to queue ops, if not set by enqueue ops.

PiperOrigin-RevId: 200764324

dimvar committed:
Typo fixes.

PiperOrigin-RevId: 200766687

Commit:
Fix None grads bug when calling a keras Sequential twice on same input in graph
mode.

PiperOrigin-RevId: 200768236

MarkDaoust committed:
Quiet the doc generator.

Delete most print statements, use logging instead of print, and close files (to clear the "Unclosed file" warnings).

Normally this produces thousands of lines of output. Mostly noise.

PiperOrigin-RevId: 200769210

tensorflower-gardener committed:
Add bazel android repo to workspace

PiperOrigin-RevId: 200771096

rryan committed:
Fix typo in tf.lite Python interpreter comment.

PiperOrigin-RevId: 200774484

kayzhu committed:
Automated g4 rollback of changelist 200750664

PiperOrigin-RevId: 200777514

skye committed:
Move cond_v2 to core (non-public) and add toggle to use cond_v2 by default.

This change:
* Creates a new global variable, control_flow_ops._ENABLE_COND_V2, to use
  cond_v2 by default when calling tf.cond. This variable can also be
  controlled via the environment variable TF_ENABLE_COND_V2.

* Moves cond_v2 out of contrib so it's accessible from control_flow_ops.py.

* Lazily "imports" some modules in cond_v2 to avoid circular dependencies.
  Note that these lazy "imports" must be imported by the cond_v2 caller (or
  recursively by one of the caller's imports) in order for cond_v2 to have
  access to them.

* Renames the cond_v2 module to cond_v2_impl, and creates a new cond_v2 module
  that imports the cond_v2 method and the necessary extra imports. This is
  useful for explicitly calling cond_v2 outside of control_flow_ops.cond.

PiperOrigin-RevId: 200778208

Commit:
Reflow comments; NFC

PiperOrigin-RevId: 200783258

reedwm committed:
Automatic cast layer inputs to the layer's dtype.

This makes it more convenient to use layer of different dtypes in a model. Instead of having to manually cast intermediate tensors between layers of different dtypes, they will automatically be casted.

This is also useful for the upcoming mixed precision API.

PiperOrigin-RevId: 200783477

skye committed:
Enable fetching shapes from the C API by default.

Prior this change, we were using the C API for everything except
Tensor.shape calls, which returned the result from the original Python
shape inference code. With this change, we use the C API in this case
as well. The C API has better shape inference, so this has the effect
of returning more precise shapes in some cases.

This change can be disabled by setting the environment variable
TF_C_API_GRAPH_CONSTRUCTION_SHAPES=0. However, this toggle will
be removed altogether in the near future.

This also fixes a bug in the SWIG that could cause large shape dimensions
to be incorrect.

PiperOrigin-RevId: 200783822

tensorflower-gardener committed:
Some fixes to testInferenceInputType

PiperOrigin-RevId: 200789288

tensorflower-gardener committed:
Split GradientBoostedDecisionTreeModel.train() to three steps.
1) Update stats
2) Update the number of examples visited.
3) If the number of examples reaches the target, grow the tree.

PiperOrigin-RevId: 200790145

saeta committed:
Integrate ClusterResolvers with Keras TPU support

PiperOrigin-RevId: 200790410

tensorflower-gardener committed:
Fixes Eager mode of dynamic_rnn for RNNCells with unbalanced output

PiperOrigin-RevId: 200791012

Commit:
Faster TopoQueue in graph_properties.

PiperOrigin-RevId: 200791799

asimshankar committed:
[py_func]: Fix #20021

* EagerPyFunc now validates its assumption that returned tensors are backed by memory on the same device that the EagerPyFunc kernel executed on.
* Make the Python trampolining mechanism ensure that this requirement of the kernel is met.
* Allow tf.contrib.eager.py_func to execute correctly on devices other than CPU and GPU:0.

Prior to this change, tf.contrib.eager.py_func() would copy data from CPU to GPU:0 if necessary, but not the other way around. As a result, the assumptions made by the EagerPyFunc kernel implementation about the placement of returned tensors would be violated.

The test added in py_func_test.py, when executed on a machine with a GPU will:

- Fail with a segmentation fault (dereferencing GPU memory) without the changes to py_func.cc and script_ops.py
- Fail with an error message with the change to py_func.cc but without the change to script_ops.py
- Pass with changes to py_func.cc and script_ops.py

PiperOrigin-RevId: 200792057

akshayka committed:
Allow Tensor-valued keyword arguments for tfe.defun.

The full list of inputs to the generated TF function is created by appending
the Tensor-valued keyword arguments (sorted by key) to the list of
Tensor-valued args.

PiperOrigin-RevId: 200792676

majnemer committed:
[XLA] Implement ConjugateTransposeOp

This simply wraps the Transpose with a Conj.

PiperOrigin-RevId: 200793274

tensorflower-gardener committed:
Expose Quantization params for outputs in JNI interpreter

PiperOrigin-RevId: 200795402

allenlavoie committed:
Add tf.contrib.checkpoint.CheckpointableBase for isinstance checks.

(Also planning to use this in Sonnet)

PiperOrigin-RevId: 200797385

tensorflower-gardener committed:
Automated g4 rollback of changelist 200790145

PiperOrigin-RevId: 200799531

jsimsa committed:
[tf.data] Concurrency improvements to `map_and_batch`.

PiperOrigin-RevId: 200800013

skye committed:
Remove bad assert in control_flow_ops.py.

TensorShape.__eq__ will return false if there are any unknown
dimensions in the shapes being compared, even if both shapes have
unknown dims in the same place. This means that the assert in
control_flow_ops.py would sometimes spuriously trigger. This change
removes the assert since it was for debugging anyway.

PiperOrigin-RevId: 200801159

iganichev committed:
TFE: Correctly set shapes of defun outputs

When a function being converted to defun runs, it can output
non-tensor values. The "shape" of these non-tensor values is
set to None in _output_shapes. When we set the shapes at the
end of the defun __call__, we need to skip these Nones.

Also, unrelatedly, add a test for basic gradient tape and a test
for using strided_slice inside a compiled and taped defun.
strided_slice "stresses" XLA's constant inference for arguments
that must be constant.

PiperOrigin-RevId: 200802717

k-w-w committed:
Automated g4 rollback of changelist 200747752

PiperOrigin-RevId: 200802842

tensorflower-gardener committed:
Don't check for duplicates in FetchOutputs and FeedInputs when creating a ControlEdge. There cannot be a duplicate, since fetch_node and feed_node are newly created. This change reduces the complexity of FetchOutputs from quadratic to linear.

PiperOrigin-RevId: 200807286

jsimsa committed:
[tf.data] Internal refactor of the parallel version of `tf.data.Dataset.map()`, switching from using a fixed-size circular buffer to a deque.

PiperOrigin-RevId: 200808498

thomasjoerg committed:
[XLA:GPU] Allow different element types in multi-output fusion root tuples.

PiperOrigin-RevId: 200809229

gunan committed:
Disable random_ops_test on windows.

PiperOrigin-RevId: 200814177

gunan committed:
Automated g4 rollback of changelist 200623983

PiperOrigin-RevId: 200817339

tensorflower-gardener committed:
TFLite Custom op for object detection postprocessing.

PiperOrigin-RevId: 200820561

Commit:
Teach gather-reshape folding to work with degenerate dims

I was hoping not to do this, but the motivating benchmark for all this work has
reshapes on degenerate dimensions.  This also forced me to introduce a new node
to the analysis which isn't great (we don't want to replicate HLO inside
IndexedArrayAnalysis!) but this is cleanest solution I can think of.

In brief I support gather-reshape folding with degenerate dimensions by
disallowing it in the core tricky part of the algorithm and instead reshaping
the degenerate dimensions "in and out" in a helper that calls the core part of
the folding logic.

Also worth calling out that before we weren't doing something conservative -- we
were just buggy.  For instance the CHECK_NE(candidate_operand_dim, 0) in
ComputeReshapePassthroughDimPairs can fail with degenerate dims.

I also made some other supporting changes:

 - I was not checking window bounds in ComputeArrayForGather.  I've fixed this
   and beefed up testing in this area (the hammer for all my nails).
 - Added a bunch of VLOG(3) info that was useful when debugging.
 - Added a simple helper to the test that makes the strings I'm matching against
   "whitespace insensitive" so that I can indent these.

I'm happy to pull these out into separate CLs if that makes reviewing easier but
for now I took the path of least resistance. :)

PiperOrigin-RevId: 200821883

tensorflower-gardener committed:
Optimize max/min reductions over monotonic functions

PiperOrigin-RevId: 200843761

Commit:
clean up

PiperOrigin-RevId: 200849332

cdleary committed:
[XLA] Propagate StatusOr through SWIG interface.

PiperOrigin-RevId: 200852741

tensorflower-gardener committed:
Convert ImportTensorFlow method from switch to table based.

PiperOrigin-RevId: 200892708

tensorflower-gardener committed:
Implement reduce_sum

PiperOrigin-RevId: 200895985

tensorflower-gardener committed:
Fix minor bug in handling of IndicatorColumn in BoostedTreesClassifier.

Handles case where max_buckets_for_bucketized overwrites an existing key in the bucket_size_to_feature_ids_dict.

This can happen if

a) There are no bucketized features
b) The max buckets for bucketized features is actually 2 (clashing with max_buckets_for_indicator)

PiperOrigin-RevId: 200908269

aselle committed:
Add support for direct buffer access from TF Lite Python API.

Also fixed other problems
- Fix bounds checking on tensor index
- Fix tensor byte size to be size_t
- Fix memory leak in buffer allocation
- Remove dependency on core tensorflow

In a susbsequent CL I will refactor to not require logging and instead send
ValueError or RuntimeErrors back as exceptions that properly use TFLite
ErrorReporters.

PiperOrigin-RevId: 200915674

gunan committed:
Disable flaky random_ops_test

PiperOrigin-RevId: 200934420

tensorflower-gardener committed:
Enable bfloat propagation for bitcast HLO

If the input and output element type for a bitcast is the same (it is
only a layout and shape change) then its effective output precision is
same as its input precision.

PiperOrigin-RevId: 200966788

tensorflower-gardener committed:
Support BF16 propagation through domain instructions

Domain instructions only there to carry some metadata so they don't
effect the precision of the data so we should propagate BF16 through
them.

The special code needed to handle domain instructions is there as this
is the only HLO what have the same tuple shaped operand and result.

PiperOrigin-RevId: 200968713

tensorflower-gardener committed:
Use --output_user_root to specify a short output base for Windows build (Prepare for upgrading Bazel to 0.14.1 on Windows)

PiperOrigin-RevId: 200988382

d0k committed:
[XLA:GPU] Don't run layout assignment (or any HLO passes) in multioutput fusion test

This allows making the GPU emitter checks more restrictive (this would be a miscompile otherwise). Layout assignment cannot run with pre-assigned layouts currently.

PiperOrigin-RevId: 200993754

tensorflower-gardener committed:
Rename object detection custom op filenames to be consistent with earlier comments on renaming the file and op.

PiperOrigin-RevId: 200999974

d0k committed:
[XLA] Change calls to LiteralTestUtil::Equal to pass in the expected value first

This makes the failure output less confusing.

PiperOrigin-RevId: 201001511

jsimsa committed:
[tf.data] Cleanup of tf.contrib.data python tests.

PiperOrigin-RevId: 201004909

tensorflower-gardener committed:
Add missing strip_prefix to workspace.

PiperOrigin-RevId: 201005676

hawkinsp committed:
[TF:XLA] Add a XlaSort operator that directly wraps the Sort HLO.

Merge XLA-specific operator registrations into a single file rather than having many tiny files.

In passing, register a fill function for bfloat16 numpy type; needed for the np.arange() call in the sort unit test.

PiperOrigin-RevId: 201005718

Commit:
Merge changes from github.

PiperOrigin-RevId: 201011811

allenlavoie committed:
Don't add to the global losses collection from tf.losses.* when executing eagerly

Fixes #20062.

RELNOTES: tf.losses.* do not add to the global collection when executing eagerly (avoids leaking memory).
PiperOrigin-RevId: 201015215

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 201015958

hawkinsp committed:
Include the name of the resource in error messages about cross-device resource access.

PiperOrigin-RevId: 201032994

Commit:
Automated g4 rollback of changelist 201011811

PiperOrigin-RevId: 201033171

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 201037916

majnemer committed:
[XLA] Simplify, add additional testing for TruncatedNormal

PiperOrigin-RevId: 201039966

tensorflower-gardener committed:
fix ReadTensor not reading the full contents of reader

PiperOrigin-RevId: 201040414

tensorflower-gardener committed:
Have TensorFlow use latest version of nsync.

There is no significant change for popular platforms, and most users will not notice.
Some unpopular platforms have better support for atomics.

PiperOrigin-RevId: 201040944

gunan committed:
Disable large tests in fastbuild mode.

PiperOrigin-RevId: 201048439

tensorflower-gardener committed:
Enables `If` operator lowering in cond_v2 when XLA is disabled. Lowering allows cond_v2 to avoid some of the limitations of Functions, allowing users to specify devices & colocation inside of cond_v2 branches, and enabling non-strict evaluation & partial pruning of branches. This brings cond_v2 closer to feature parity with tf.cond.

However, we do not lower `If` in the XLA context because it is easier for XLA to apply its own optimizations when dealing with un-lowered `If` operators than with lowered switch/merge control flow.

Also adds a toggleable flag in for InlineFunctionBody in function.cc that prevents the function caller device from overriding the devices of function body nodes. This is necessary for cond_v2 branches to support explicitly-specified devices.

Adds several tests to make sure that:
- lowering is usually enabled
- lowering is disabled for XLA
- node colocation inside of cond_v2 branches works
- explicit device placement inside of cond_v2 branches works

PiperOrigin-RevId: 201049850

d0k committed:
[XLA:GPU] Un-unimplement gather emission

We already have elemental code for doing this in the fused case, this just
enables it in the unfused case.

PiperOrigin-RevId: 201050143

tensorflower-gardener committed:
Migrate to android skylark rules

PiperOrigin-RevId: 201052263

Commit:
Broad refactoring (part 3): reorganize the code so that the dependency graph is cleaner and better separates AutoGraph logic from the general purpose SCT code (concentrated in pyct).

The new module structure is described in CONTRIBUTING.md.

Summary of changes:
 * the new lang and core modules now replace their old counterparts
 * CONTRIBUTING.md now has a short paragraph on developer info
 * the lang APIs are exposed into the main autograph interface
 * the old implementations for converter_test_base.py, config.py, directives.py, naming.py, special_functions.py and their tests are now removed
 * all converters now inherit converter.Base instead of transformer.Base
 * all converter tests now inherit converter_testing.TestCase instead of converter_test_base.TestCase
 * converter interfaces now all share a common signature: <converter>.transform(node, context)
 * the decorator module now actually imports dependencies requires for existing decorators, which was previously just a TODO
 * decorator_test now runs an additional test that was previously disabled
 * the implementation of conversion.node_to_graph is now simpler and more consistent; ConversionMap is removed
 * type_info.py now creates a separate "definition" annotation for all symbols
 * transformer.py no longer has any mention to AutoGraph specific implementations
 * other no-op code simplifications, doc and comment updates

PiperOrigin-RevId: 201053048

tensorflower-gardener committed:
Convert CheckInputsSize to return a Status instead of CHECK-failing, and convert existing callsites to TF_QCHECK_OK the call.

This moves us towards the goal of returning Statuses instead of check-failing in ImportTensorFlowNode().

PiperOrigin-RevId: 201056489

tensorflower-gardener committed:
Replace distribution_util.assert_close with tf.assert_near.

PiperOrigin-RevId: 201058937

tensorflower-gardener committed:
This is an initial submission of GGT to tensorflow contrib.
Paper link: https://arxiv.org/pdf/1806.02958.pdf

PiperOrigin-RevId: 201063723

tensorflower-gardener committed:
Support rsqrt for graphdef export.

PiperOrigin-RevId: 201067685

tensorflower-gardener committed:
Fix bug in RemoveIdempotent optimizer stage.
Minor cleanup in RemoveIdentityTranspose.

PiperOrigin-RevId: 201069367

Commit:
Fix input_batch_size for PER_HOST_V2 when model parallelism is enabled.

PiperOrigin-RevId: 201070853

tensorflower-gardener committed:
Enable the natural layouts of the entry computation to flow into the parameters and result layouts of the entry ComputationLayout.
If the arguments shapes passed in to the servie.cc API do not have a layout, it is assumed the caller is willing to accept the natural layout propagated by the XLA compiler.
Similarly, if the ExecutionOptions has a shape for the result, but no layout is set in such shape, it is assumed the caller is willing to accept the natural layout propagated by the XLA compiler.
Same thing for the ExecutableBuildOptions result_layout().

PiperOrigin-RevId: 201070858

tensorflower-gardener committed:
Make sure CRS is fully deserializable from an HLO TXT and Proto POV.

PiperOrigin-RevId: 201070859

asimshankar committed:
Java: Release 1.9.0-rc1

PiperOrigin-RevId: 201071358

iganichev committed:
Default to compiling functions running on TPU.

PiperOrigin-RevId: 201071433

tensorflower-gardener committed:
Extract tf_record_test.py from reader_ops_test.py

PiperOrigin-RevId: 201071448

tensorflower-gardener committed:
Internal change.

PiperOrigin-RevId: 201073792

ispirmustafa committed:
Improving local run behavior in estimator.train_and_evaluate.
Current behavior is unintuitive (depends on throttle_secs) and leads to frequent checkpoint than desired.
This CL makes evaluation synchronized with checkpointing. It also makes the behavior more closer to distributed setting in following ways:
* in distributed setting we do create input_pipeline only once, in current behavior of local run we do recreate input pipeline in a loop. This cl creates training input pipeline only once.
* in distributed setting evaluator job waits for checkpoints which are dumped by training job. In current behavior of local run evaluator controls the checkpoint schedule. In this cl, we give back the control to trainer.

PiperOrigin-RevId: 201085814

shashishekhar committed:
Add an iOS benchmarking app.

PiperOrigin-RevId: 201085939

tensorflower-gardener committed:
The pretrained text embedding models in tf.hub expect a string input. If I pass dtype as tf.string in tf.keras.layers.InputLayer, it fails in a numpy array conversion as numpy doesn't recognize tf string type. I have added a check for that and if the input is a string, then the dtype passed to np.asarray is object.

PiperOrigin-RevId: 201085946

Commit:
Make learning decay functions return functions that return the
learning rate tensor.

This enables proper learning rate schedules in eager mode.

PiperOrigin-RevId: 201089859

tensorflower-gardener committed:
Make NNAPI delegation support more ops.

PiperOrigin-RevId: 201090056

tensorflower-gardener committed:
Split out opcodes with window as subclasses from HloInstruction (kConvolution, kReduceWindow, kSelectAndScatter, kCustomCall).

PiperOrigin-RevId: 201093426

tensorflower-gardener committed:
Migration to python 3 for estimator.predict.

PiperOrigin-RevId: 201093768

tensorflower-gardener committed:
Reduce Grappler overhead by skipping optimizers when the graph is tiny.

PiperOrigin-RevId: 201095811

gunan committed:
Add missing numpy header dependency to pywrap_tfe_lib

PiperOrigin-RevId: 201095991

hawkinsp committed:
[TF:XLA] Implement TopKV2 for bfloat16 types by packing into a float32

PiperOrigin-RevId: 201100290

tensorflower-gardener committed:
HloInstruction::CreateFromProto should not crash on CHECK, instead needs to return error status.
PiperOrigin-RevId: 201100918

Commit:
Automated g4 rollback of changelist 201089859

PiperOrigin-RevId: 201101839

tensorflower-gardener committed:
Split out HloHostComputeInstruction, HloPadInstruction and HloDynamicSliceInstruction as subclasses from HloInstruction..

PiperOrigin-RevId: 201108336

smit-hinsu committed:
Load NCCL lib on-demand to facilitate default NCCL version upgrade to 2

Change in the default version to NCCL 2 would require all TF users to
download the NCCL library without the on-demand loading. With on-demand
loading, it will only require users using the nccl ops to download and
install the NCCL lib.

PiperOrigin-RevId: 201109554

Commit:
Merge changes from github.

PiperOrigin-RevId: 201110240

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 201111838

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 201113951

gunan committed:
Automated g4 rollback of changelist 200988382

PiperOrigin-RevId: 201119398

ilya-biryukov committed:
Update downloadable clang to r334100.

PiperOrigin-RevId: 201127564

akuegel committed:
Mark Gather as fusile.

There is an elementwise implementation for Gather.

PiperOrigin-RevId: 201136554

tensorflower-gardener committed:
Derivative of tf.igamma(a, x) and tf.igammac(a, x) with respect to a.

Previously, both functions only supported the derivative with respect to x. We add the derivative with respect to the other argument. It is computed using the Eigen function igamma_der_a that performs forward-mode differentiation of the code for igamma. This function is not exposed in the public TensorFlow API.

PiperOrigin-RevId: 201145398

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 201147873

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 201150427

protoget committed:
Update a few documentation for layer-input-casting feature.

PiperOrigin-RevId: 201152785

tensorflower-gardener committed:
Internal change

PiperOrigin-RevId: 201161803

Commit:
Disable test on windows.

PiperOrigin-RevId: 201163760

tensorflower-gardener committed:
Performance microtweaks: Pass by reference rather than by value; pre-reserve capacity when total vectoroid size is known.

PiperOrigin-RevId: 201172723

mrry committed:
[tf.data] Fix a performance-related finding from clang-tidy.

* the parameter 'done' is copied for each invocation but only used as a const reference; consider making it a const reference

PiperOrigin-RevId: 201179686

tensorflower-gardener committed:
Update TFLite "minimal" example

PiperOrigin-RevId: 201183828

Commit:
Automated g4 rollback of changelist 201069367

PiperOrigin-RevId: 201190626

hawkinsp committed:
[TF:XLA] Add a global mutex around XlaCompileOnDemandOp's call to Executable::Run() to work around a concurrency problem in XLA.

PiperOrigin-RevId: 201191495

tensorflower-gardener committed:
Implement TFLite sqrt/rsqrt unary operators

PiperOrigin-RevId: 201191877

tensorflower-gardener committed:
Refresh allocations in the presence of dynamic tensors

PiperOrigin-RevId: 201193941

tensorflower-gardener committed:
Apply runtime shapes to pooling and activation kernels.

PiperOrigin-RevId: 201194552

tensorflower-gardener committed:
Use --output_user_root to specify a short output base for Windows build (Prepare for upgrading Bazel to 0.14.1 on Windows)

PiperOrigin-RevId: 201197774

Commit:
Allow setting server def on the eager context, and add the eager service to the grpc_tensorflow_server.

PiperOrigin-RevId: 201198350

tensorflower-gardener committed:
Automated g4 rollback of changelist 201190626

PiperOrigin-RevId: 201202998

tensorflower-gardener committed:
Adding reference to the following classes:
ConvolutionDeltaOrthogonal
ConvolutionOrthogonal1D
ConvolutionOrthogonal2D
ConvolutionOrthogonal3D

PiperOrigin-RevId: 201203440

reedwm committed:
Automated g4 rollback of changelist 200783477

PiperOrigin-RevId: 201204573

akshayka committed:
Correctly compute real and side outputs when constructing backprop function.

Prior to this change, we assumed that the number of real outputs of the TF
function was equal to the number of outputs of the Python function. This
assumption was incorrect, as the Python function might return non-Tensor
objects whereas the TF function exclusively returns Tensors.

PiperOrigin-RevId: 201205657

Commit:
Internal Change.

PiperOrigin-RevId: 201208955

tensorflower-gardener committed:
Reconcile enum types.

PiperOrigin-RevId: 201210730

Commit:
[TF:XLA] Bump open source llvm revision to r335024

PiperOrigin-RevId: 201213520

tensorflower-gardener committed:
Rollback of changelist 200200356. We might want to support GPUs on MacOS again in the future. Users are interested to make it work and we don't want to be in the way.

PiperOrigin-RevId: 201214857

tensorflower-gardener committed:
fix a bug about converting Log1p - we are checking the x tensor (not the constant tensor) to be 1.

PiperOrigin-RevId: 201217989

reedwm committed:
Rollback documentation that I forgot to rollback last time.

PiperOrigin-RevId: 201218249

miaout17 committed:
Support Variable Tensor API in LSTM Full kernel.

TFLite LSTM now supports 5 inputs, 18 inputs and 20 inputs.

PiperOrigin-RevId: 201222516

jsimsa committed:
Changing test size to "medium" to prevent test timeouts.

PiperOrigin-RevId: 201225326

rohan100jain committed:
Moving SharedEmbeddingColumns state management back to graph collections. Erroring out SharedEmbeddingColumn usage in Eager mode since collections aren't supported in eager.

PiperOrigin-RevId: 201230316

tensorflower-gardener committed:
Disable caching_device for mirrored variables.

PiperOrigin-RevId: 201232817

Commit:
Automated g4 rollback of changelist 201101839

PiperOrigin-RevId: 201236075

tensorflower-gardener committed:
Refactor the impl of Shard() so that the caller can use a Runner.

PiperOrigin-RevId: 201236564

guptapriya committed:
Create hyper parameter tensors in optimizer v2 outside any control flow contexts.
Also, use lambdas for creating the non slot variables in adam v2.
These changes are needed to allow optimizer.minimize to run inside a while loop, which will be done in distribution strategies shortly.

PiperOrigin-RevId: 201238566

tensorflower-gardener committed:
Improve filter for cuBLAS bug.

PiperOrigin-RevId: 201239428

tensorflower-gardener committed:
Automated g4 rollback of changelist 201194552

PiperOrigin-RevId: 201241214

Commit:
Rename llvm.BUILD to llvm.autogenerated.BUILD

In practice folks tend to miss the "# This BUILD file is auto-generated; do not
edit!" admonition.

PiperOrigin-RevId: 201248010

Commit:
Allow default TF/XLA op registration with specific backend overrides.

PiperOrigin-RevId: 201252399

kayzhu committed:
Automated g4 rollback of changelist 200777514

PiperOrigin-RevId: 201252470

tensorflower-gardener committed:
ConfigureGcsHooks: Fixed a couple of typos.

- _configure_op was spelled with a trailing 's'
- _block_cache_op was only conditionally set but unconditionally read.

Added a fake test that triggered the bugs before and passes after.

PiperOrigin-RevId: 201256874

dimvar committed:
[TF:XLA] Prevent overflow in hlo_scheduling, when compiling AutoML models.

PiperOrigin-RevId: 201257475

tanzhenyu committed:
Add an advanced activation layer for ReLU

PiperOrigin-RevId: 201257601

tensorflower-gardener committed:
Automated g4 rollback of changelist 201217989

PiperOrigin-RevId: 201257755

tensorflower-gardener committed:
Derivative of tf.random_gamma with respect to the alpha parameter.

Previously, tf.random_gamma(shape, alpha, beta) was differentiable only w.r.t. beta. This commit adds the derivative w.r.t. alpha. The implementation is based on Eigen's gamma_sample_der_alpha function, which computes the "implicit reparameterization" derivative. This function is not directly exposed in the public TensorFlow API.

PiperOrigin-RevId: 201258617

Commit:
Add scripts to write to tfrecords, read from tfrecords and training.

PiperOrigin-RevId: 201263223

k-w-w committed:
Refactor loader.load function into a class that splits the graph loading and variable restoration steps.

PiperOrigin-RevId: 201268712

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 201269772

rohan100jain committed:
Fixing a bug in linear_model where the name for the model is always set to 'linear_model'. This causes issues when we create multiple linear models in the same graph.

PiperOrigin-RevId: 201270816

broune committed:
Add interface in Compiler for computing the default backend configuration of an op.

Add interface in Executable for computing the size of the executable.

PiperOrigin-RevId: 201271132

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 201273382

tensorflower-gardener committed:
[TF:XLA] Fix for HLO instruction post-order DFS and multioutput fusion.

Cycles were not handled correctly when computing the postorder of an HLO computation.

Add methods to multioutput fusion that allows subclasses to recompute and query the
current reachability map.

PiperOrigin-RevId: 201274181

guptapriya committed:
Implement new API for TPUStrategy to run multiple steps, and move most of the TPU specific logic into this method from `call_for_each_tower`.
Disable TPU tests temporarily, will enable again in subsequent code changes.

PiperOrigin-RevId: 201279470

tensorflower-gardener committed:
Remove the ambiguity of device/host computation layouts within the HloModuleConfig.

PiperOrigin-RevId: 201284741

majnemer committed:
[TF2XLA] Optimize TruncatedNormalOp

Re-sampling when encountering a rejected value can be quite slow.
If we directly use the inverse CDF of the normal distribution, the probit
function, we can avoid the need to resample.

PiperOrigin-RevId: 201296864

tensorflower-gardener committed:
Internal change

PiperOrigin-RevId: 201301504

lamberta committed:
Get started landing page.
Move "Datasets Quickstart" to "Datasets for Estimators" under guide.

PiperOrigin-RevId: 201301717

bixia1 committed:
Allow the use of 32 bit integer type for loop index and tensor element index.

The GPU LLVM IR generator currently uses 64 bit integer type for arithmetic
operations related to loop index and tensor element index and relies on LLVM
optimization to narrow the operations to 32 bit integer type. There are
situations whether LLVM optimization fail to perform such an optimization, see
LLVM D46760 for more detail.

This change modifies the XLA LLVM IR code generation infrastructure to support
the use of 32 bit integer type for loop index and tensor element index as
follows:
 .Extends the loop emitter interface in ParallelLoopEmitter and ForLoopNest to
  allow users to specify the loop index type.
 .Modifies the tensor access interface in IrArray::Index interface to record
  the llvm type for the index when an object is constructed. This index type
  is usually propagated from a loop index type.
 .Modifies kernel_support_library to retrieve the loop index type from the
  input llvm::Value.
 .Modifies elemental_ir_emitter to retrieve the data type from the input
  IrArray::Index and use it tensor offset expression.

This change also modifies the emission of the fusion kernel, the row and
scalar reduction kernel and SelectAndScatter kernel to use 32 bit integer type
for index calculation when the size of the launch dimension and the size of
tensors used in the kernel are within the range of 32 bit integer
representation.

PiperOrigin-RevId: 201303468

tensorflower-gardener committed:
Improvements in the documentation of tf.random_gamma, tf.random_poisson and tf.distributions.

* Marked the Python code in docstrings.
* Fixed the output shapes in docstrings.
* Fixed a typo in the normalization constant in tf.distributions.Gamma docstring.
* Updated the warning in tf.distributions.Gamma docstring.
* Added warnings regarding zero samples in tf.distributions.Beta and tf.distributions.Dirichlet docstrings.

PiperOrigin-RevId: 201328305

akuegel committed:
[TF:XLA] Bump open source llvm revision to r335074

PiperOrigin-RevId: 201337140

akuegel committed:
Simplify ConvertLiteralToIrConstant()

Also use ConstantDataArray for C64 types.
This allows to delete the old LiteralToDataConstant() method.

PiperOrigin-RevId: 201339634

tensorflower-gardener committed:
Make common_runtime/eager libraries compile for Android, by eliding the dependency on
GRPC.

PiperOrigin-RevId: 201353152

tensorflower-gardener committed:
Support list of integers in custom op attributes.

PiperOrigin-RevId: 201356549

akshayka committed:
Support defun-ing instance methods.

This change implements the __get__ method on _PolymorphicFunction and has
it forward the instance to __call__. This makes it possible to write code like

class Foo(object):
  ...

  @tfe.defun
  def two(self, tensor):
    ...

PiperOrigin-RevId: 201365344

lamberta committed:
Fix eager path in get_started leftnav

PiperOrigin-RevId: 201370156

tensorflower-gardener committed:
Fix a bug that would leave orphaned arrays in the graph.

PiperOrigin-RevId: 201370219

mrry committed:
[tf.data] Properly export `tf.contrib.data.choose_from_datasets()`

PiperOrigin-RevId: 201371642

asimshankar committed:
[eager]: Support string attributes where the value contains `\0`.

Apparently, some custom operations stuff non-printable characters in string
valued attributes.

This change also makes the eager C API consistent with the C API for graph
construction (TF_SetAttrString and TF_SetAttrStringList).

PiperOrigin-RevId: 201372089

asimshankar committed:
s/tf.contrib.eager.GradientTape/tf.GradientTape/

PiperOrigin-RevId: 201372249

MarkDaoust committed:
Hide py3 names we don't need to document.

PiperOrigin-RevId: 201374225

tensorflower-gardener committed:
16-bit quantized add support in TFLite interpreter

PiperOrigin-RevId: 201374318

Commit:
Allow TowerLocalVars to be updated with the same value across all towers.

PiperOrigin-RevId: 201379124

dimvar committed:
[TF:XLA] Change hlo_domain_test to use HloVerifiedTestBase.

PiperOrigin-RevId: 201383246

gunan committed:
Update protobuf dependency of TF to 3.6.

PiperOrigin-RevId: 201386306

jsimsa committed:
[tf.data] Properly export `tf.contrib.data.group_by_reducer()`

PiperOrigin-RevId: 201386380

Commit:
Boilerplate for an ANF transformer. This is not currently related to AutoGraph, but used elsewhere.

PiperOrigin-RevId: 201387308

jdduke committed:
Implement TFLite Shape operator

PiperOrigin-RevId: 201389618

tensorflower-gardener committed:
Fix a bug: the conversion of pure Conv to DepthwiseConv did not
properly check the necessary precondition that the input depth is 1.

PiperOrigin-RevId: 201389819

tensorflower-gardener committed:
Make evaluate() work on anything that has a numpy() method in eager tests.

PiperOrigin-RevId: 201390698

alextp committed:
SymbolicGradient for some resource variables.

Currently assumes variables are floats; there are TODOs to rectifiy this.

PiperOrigin-RevId: 201391092

shashishekhar committed:
Fix gradle build for TFLite Android example.

PiperOrigin-RevId: 201391220

tensorflower-gardener committed:
Split GradientBoostedDecisionTreeModel.train() to three steps.
1) Update stats
2) Update the number of examples visited.
3) If the number of examples reaches the target, grow the tree.

PiperOrigin-RevId: 201392512

mingxingtan committed:
Merge commit for internal changes

mingxingtan committed:
Fix minor merging issue.

mingxingtan committed:
Fix an XLA merging error.

yifeif commented:
@mdanatg do you mind taking a look at the tensorflow/contrib/autograph/converters:decorators_test failure?

mdanatg commented:
@yifeif Looking now. Looks like a missing dependency that somehow I missed.

mingxingtan commented:
Thanks @mdanatg and @yifeif 
@mdanatg, this is blocking our push, could you prioritize it and keep me posted? Thanks!

yifeif committed:
Disable decorators_test for pip.

tensorflow/contrib/autograph/converters:decorators_test uses generated code, by when private symbols have been stripped.
